File: src/main/java/io/github/gdpl2112/UserController.java
package io.github.gdpl2112;

import io.github.gdpl2112.model.StorageInfo;
import io.github.gdpl2112.model.User;
import io.github.gdpl2112.model.UserFile;
import io.github.gdpl2112.service.AuthService;
import io.github.gdpl2112.service.FileService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import jakarta.servlet.http.HttpSession;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.List;

@RestController
@RequestMapping("/user")
public class UserController {

    @Autowired
    private AuthService authService;

    @Autowired
    private FileService fileService;

    @Value("${file.upload-dir}")
    private String uploadDir;

    /**
     * 获取用户信息
     *
     * @param session HTTP会话
     * @return 用户信息
     */
    @GetMapping("/info")
    public ResponseEntity<User> getUserInfo(HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user != null) {
            return ResponseEntity.ok(user);
        } else {
            return ResponseEntity.status(401).build();
        }
    }

    /**
     * 获取存储使用情况
     *
     * @param session HTTP会话
     * @return 存储信息
     */
    @GetMapping("/storage")
    public ResponseEntity<StorageInfo> getStorageInfo(HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        StorageInfo storageInfo = new StorageInfo();
        long limit = user.getStorageLimit();
        long used = fileService.calculateFolderSize(user);
        long remaining = limit - used;
        double percentage = limit > 0 ? (double) used / limit * 100 : 0;

        storageInfo.setLimit(limit);
        storageInfo.setUsed(used);
        storageInfo.setRemaining(remaining);
        storageInfo.setPercentage(percentage);
        storageInfo.setLimitFormatted(StorageInfo.formatFileSize(limit));
        storageInfo.setUsedFormatted(StorageInfo.formatFileSize(used));
        storageInfo.setRemainingFormatted(StorageInfo.formatFileSize(remaining));

        return ResponseEntity.ok(storageInfo);
    }

    /**
     * 获取用户文件列表
     *
     * @param session HTTP会话
     * @return 文件列表
     */
    @GetMapping("/files")
    public ResponseEntity<List<UserFile>> getUserFiles(HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        List<UserFile> files = fileService.getUserFiles(user);
        return ResponseEntity.ok(files);
    }

    /**
     * 检查文件是否存在
     *
     * @param path 文件路径
     * @return 是否存在
     */
    @GetMapping("/exits")
    public ResponseEntity<Boolean> checkFileExists(@RequestParam String path,HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            return ResponseEntity.status(401).build();
        }
        File file = new File(user.getPathDir(uploadDir), path);
        return ResponseEntity.ok(file.exists());
    }

    /**
     * 下载文件
     *
     * @param filename 文件名
     * @return 文件资源
     */
    @GetMapping("/download/{filename:.+}")
    public ResponseEntity<Resource> downloadFile(HttpSession session, @PathVariable String filename) {
        User user = (User) session.getAttribute("user");
        if (user == null) return ResponseEntity.status(401).build();
        Path file = Paths.get(user.getPathDir(uploadDir)).resolve(filename);
        try {
            Resource resource = new UrlResource(file.toUri());
            if (resource.exists() || resource.isReadable()) {
                return ResponseEntity.ok()
                        .header(HttpHeaders.CONTENT_DISPOSITION, "attachment; filename=\"" + resource.getFilename() + "\"")
                        .body(resource);
            } else {
                return ResponseEntity.notFound().build();
            }
        } catch (IOException e) {
            return ResponseEntity.internalServerError().build();
        }
    }

    // 删除
    @DeleteMapping("/delete/{filename:.+}")
    public ResponseEntity<String> deleteFile(@PathVariable String filename, HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            return ResponseEntity.status(401).body("未登录");
        }

        File file = new File(user.getPathDir(uploadDir), filename);
        if (!file.exists()) {
            return ResponseEntity.notFound().build();
        }

        if (!file.delete()) {
            return ResponseEntity.status(500).body("删除文件失败");
        }

        return ResponseEntity.ok("删除成功");
    }

    /**
     * 上传文件
     *
     * @param file    上传的文件
     * @param session HTTP会话
     * @return 上传结果
     */
    @PostMapping("/upload")
    public ResponseEntity<String> uploadFile(@RequestParam("file") MultipartFile file, HttpSession session) {
        User user = (User) session.getAttribute("user");
        if (user == null) {
            return ResponseEntity.status(401).body("未登录");
        }

        if (file.isEmpty()) {
            return ResponseEntity.badRequest().body("文件为空");
        }

        try {
            // 检查文件大小是否超过用户存储限制
            long fileSize = file.getSize();
            long usedStorage = user.getUsedStorage();
            long storageLimit = user.getStorageLimit();

            if (usedStorage + fileSize > storageLimit) {
                return ResponseEntity.badRequest().body("存储空间不足");
            }

            // 保存文件到用户目录（这里简化处理，实际可能需要按用户ID分目录）
            File dir = new File(user.getPathDir(uploadDir));
            if (!dir.exists()) {
                dir.mkdirs();
            }
            String fileName = file.getOriginalFilename();
            if (fileName == null || fileName.isEmpty()) {
                fileName = System.currentTimeMillis() + ".dat";
            }

            File dest = new File(dir, fileName);
            byte[] bytes = file.getBytes();
            Files.write(dest.toPath(), bytes);

            // 更新用户已使用存储空间
            authService.updateUserStorage(user.getUserId(), fileSize);

            return ResponseEntity.ok("文件上传成功");
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body("上传失败: " + e.getMessage());
        }
    }
}

File: gofiles/controller/UserController.go
package controller

import (
	"net/http"
	"os"
	"path/filepath"

	"github.com/gdpl2112/files-server/model"
	"github.com/gdpl2112/files-server/service"
	"github.com/gin-gonic/gin"
)

// UserController 用户控制器
type UserController struct {
	AuthService *service.AuthService
	FileService *service.FileService
	UploadDir   string
}

// NewUserController 创建用户控制器实例
func NewUserController(authService *service.AuthService, fileService *service.FileService, uploadDir string) *UserController {
	return &UserController{
		AuthService: authService,
		FileService: fileService,
		UploadDir:   uploadDir,
	}
}

// Register 注册路由
func (c *UserController) Register(router *gin.Engine) {
	userGroup := router.Group("/user")
	{
		userGroup.GET("/info", c.GetUserInfo)
		userGroup.GET("/storage", c.GetStorageInfo)
		userGroup.GET("/files", c.GetUserFiles)
		userGroup.GET("/exits", c.CheckFileExists)
		userGroup.GET("/download/:filename", c.DownloadFile)
		userGroup.DELETE("/delete/:filename", c.DeleteFile)
		userGroup.POST("/upload", c.UploadFile)
	}
}

// GetUserInfo 获取用户信息
func (c *UserController) GetUserInfo(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	ctx.JSON(http.StatusOK, currentUser)
}

// GetStorageInfo 获取存储使用情况
func (c *UserController) GetStorageInfo(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	used := c.FileService.CalculateUserFolderSize(currentUser)
	limit := currentUser.StorageLimit
	remaining := limit - used
	percentage := 0.0
	if limit > 0 {
		percentage = float64(used) / float64(limit) * 100
	}

	storageInfo := &model.StorageInfo{
		Limit:              limit,
		Used:               used,
		Remaining:          remaining,
		Percentage:         percentage,
		LimitFormatted:     model.FormatFileSize(limit),
		UsedFormatted:      model.FormatFileSize(used),
		RemainingFormatted: model.FormatFileSize(remaining),
	}

	ctx.JSON(http.StatusOK, storageInfo)
}

// GetUserFiles 获取用户文件列表
func (c *UserController) GetUserFiles(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	files := c.FileService.GetUserFiles(currentUser)
	ctx.JSON(http.StatusOK, files)
}

// CheckFileExists 检查文件是否存在
func (c *UserController) CheckFileExists(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	path := ctx.Query("path")
	filePath := filepath.Join(currentUser.GetPathDir(c.UploadDir), path)

	exists = false
	if _, err := os.Stat(filePath); err == nil {
		exists = true
	}

	ctx.JSON(http.StatusOK, exists)
}

// DownloadFile 下载文件
func (c *UserController) DownloadFile(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	filename := ctx.Param("filename")
	filePath := filepath.Join(currentUser.GetPathDir(c.UploadDir), filename)

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		ctx.JSON(http.StatusNotFound, gin.H{"message": "文件不存在"})
		return
	}

	ctx.Header("Content-Description", "File Transfer")
	ctx.Header("Content-Disposition", "attachment; filename="+filepath.Base(filePath))
	ctx.Header("Content-Type", "application/octet-stream")
	ctx.Header("Content-Transfer-Encoding", "binary")
	ctx.Header("Expires", "0")
	ctx.Header("Cache-Control", "must-revalidate")
	ctx.Header("Pragma", "public")
	ctx.File(filePath)
}

// DeleteFile 删除文件
func (c *UserController) DeleteFile(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	filename := ctx.Param("filename")
	filePath := filepath.Join(currentUser.GetPathDir(c.UploadDir), filename)

	if _, err := os.Stat(filePath); os.IsNotExist(err) {
		ctx.JSON(http.StatusNotFound, gin.H{"message": "文件不存在"})
		return
	}

	if err := os.Remove(filePath); err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "删除文件失败: " + err.Error()})
		return
	}

	ctx.JSON(http.StatusOK, gin.H{"message": "删除成功"})
}

// UploadFile 上传文件
func (c *UserController) UploadFile(ctx *gin.Context) {
	user, exists := ctx.Get("user")
	if !exists {
		ctx.JSON(http.StatusUnauthorized, gin.H{"message": "未登录"})
		return
	}

	currentUser, ok := user.(*model.User)
	if !ok {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "用户信息类型错误"})
		return
	}

	file, err := ctx.FormFile("file")
	if err != nil {
		ctx.JSON(http.StatusBadRequest, gin.H{"message": "文件为空"})
		return
	}

	// 检查文件大小是否超过用户存储限制
	fileSize := file.Size
	usedStorage := currentUser.UsedStorage
	storageLimit := currentUser.StorageLimit

	if usedStorage+fileSize > storageLimit {
		ctx.JSON(http.StatusBadRequest, gin.H{"message": "存储空间不足"})
		return
	}

	// 保存文件到用户目录
	dir := currentUser.GetPathDir(c.UploadDir)
	if err := os.MkdirAll(dir, 0755); err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "创建目录失败: " + err.Error()})
		return
	}

	fileName := file.Filename
	if fileName == "" {
		fileName = "file.dat"
	}

	dest := filepath.Join(dir, fileName)
	if err := ctx.SaveUploadedFile(file, dest); err != nil {
		ctx.JSON(http.StatusInternalServerError, gin.H{"message": "上传失败: " + err.Error()})
		return
	}

	// 更新用户已使用存储空间
	c.AuthService.UpdateUserStorage(currentUser.UserID, fileSize)

	ctx.JSON(http.StatusOK, gin.H{"message": "文件上传成功"})
}
